### PR TITLE
Fix the behavior of the getter of VariableFeatures and how 'nfeatures =' behaves

### PR DESCRIPTION
**Pull request by longmanz**
This is a fix (with help from @Gesmira and @dcollins15 ) building on:
- https://github.com/satijalab/seurat-object/pull/213

Basically, the `VariableFeatures()` getter is re-written (using the solution provided in post 2 above) and now it respects its setter. With the this PR, users can correctly set and reset the `VariableFeatures()` (i.e., with a vector of gene names) **before** and **after** they run `FindVariableFeatures()`. The getter will now correctly return the user-defined variable features. Previously, it will return the wrong variable features (calculated by the `FindVariableFeatures()` if that has been run), as reported in post 3.

In addition, now the getter also respects the `'nfeatures = n'` flag. If `nfeatures` is set to a number smaller than the length of the existing variable features, it will output the top `nfeatures` genes, while if it is set to a number larger than the length of the existing variable features, it will output the complete list. Previously, for user-defined variable features (without running FindVariableFeatures() prior), the getter will automatically return all the VariableFeatures even when `nfeatures =` is set.

### CODE REVIEW

File: R/assay5.R [hunk 1]
```diff
@@ -1545,19 +1545,23 @@ VariableFeatures.StdAssay <- function(
...
+      }
+      if (isTRUE(x = simplify) & (is.null(x = layer) || any(is.na(x = layer))) &
+          (is.infinite(x = nfeatures) || length(x = var.features) <=
+           nfeatures)) {
+          return(var.features) 
+      } else if (isTRUE(x = simplify) & (is.null(x = layer) || any(is.na(x = layer))) &
+                 (is.infinite(x = nfeatures) || length(x = var.features) >
+                  nfeatures)) {
+          return(var.features[1:nfeatures])
+      }
```
Comment (dcollins15) [inline R/assay5.R:1564]: 🔥

**Review by dcollins15 — CHANGES_REQUESTED** [review]
Thanks for this @longmanz—looking great so far! The most important thing is that the bug seems to have been resolved 🚀 

I think we should still introduce some tests for the `VariableFeatures` getter/setter to convince ourselves that things are behaving as intended. Once we're confident in the behavior things, I think we can probably clean both methods up to make things easier to follow (and maybe more efficient). 

I've tried to get the tests started for ya:

```R
#' Returns a random counts matrix.
get_assay <- function(assay_version) {
  # Use the `assay_version` param to choose the correct assay builder.
  create_assay <- switch(assay_version,
    v3 = CreateAssayObject,
    v5 = CreateAssay5Object,
    stop("`assay_version` should be one of 'v3', 'v5'")
  )

  # Populate a 100 by 100 matrix with random integers from 1 to 50.
  counts <- matrix(
    data = sample(c(1:50), size = 1e4, replace = TRUE),
    ncol = 100,
    nrow = 100
  )

  # Assign column and row names to the matrix to label cells and genes.
  colnames(counts) <- paste0("cell", seq(ncol(counts)))
  row.names(counts) <- paste0("gene", seq(nrow(counts)))

  # Convert `counts` to a `dgCMatrix`.
  counts_sparse <- as.sparse(counts)

  assay <- create_assay(counts_sparse)

  return(assay)
}

# Adds "highly variable feature" annotations to the feature-level metadata of `assay`.
add_hvf_info <- function(assay, variable_features, method_name, layer_name) {
  all_features <- rownames(assay)
  
  # Add a column to the assay's feature-level metadata indicating which
  # features are variable.
  is_variable <- all_features %in% variable_features
  names(is_variable) <- all_features
  variable_column <- paste("vf", method_name, layer_name, "variable", sep = "_")
  assay[[variable_column]] <- is_variable

  # Add a column to the assay's feature-level metadata indicating the order
  # (i.e. "rank") that each variable feature was selected in.
  feature_rank <- seq_along(variable_features)
  names(feature_rank) <- variable_features
  rank_column <- paste("vf", method_name, layer_name, "rank", sep = "_")
  assay[[rank_column]] <- feature_rank

  return(assay)
}


test_that("`VariableFeatures` works with `StdAssay` inputs", {
  test_case <- get_assay("v5")
  
  # Generate 3 sets of "variable" features by randomly sampling 10 each.
  variable_feature_sets <- replicate(
    3,
    sample(
      rownames(test_case), 
      size = 10, 
      replace = FALSE
    ),
    simplify = FALSE
  )
  
  # Check that the getter correctly parses method-specific metadata.
  test_case <- add_hvf_info(
    test_case, 
    variable_features = variable_feature_sets[[1]],
    method_name = "methodA", 
    layer_name = "counts"
  )
  expect_identical(
    variable_feature_sets[[1]],
    VariableFeatures(test_case)
  )
  
  # Check that the setter can be used to override method-specific metadata.
  VariableFeatures(test_case) <- variable_feature_sets[[2]]
  expect_identical(
    variable_feature_sets[[2]],
    VariableFeatures(test_case)
  )
  
  # Check that the setter does not clobber the method-specific metadata.
  test_case[["var.features"]] <- NULL
  test_case[["var.features.rank"]] <- NULL
  expect_identical(
    variable_feature_sets[[1]],
    VariableFeatures(test_case)
  )
  
  # Check that the getter returns values from the most recent method.
  test_case <- add_hvf_info(
    test_case, 
    variable_features = variable_feature_sets[[3]],
    method_name = "methodB", 
    layer_name = "counts"
  )
  expect_identical(
    variable_feature_sets[[3]],
    VariableFeatures(test_case)
  )
})

```

I imagine these living in a new test module: `tests/testthat/test_variable_features.R`. This isn't a complete set, we should make sure we're also checking:

- `nfeatures` after running the `VariableFeatures` setter and the `add_hvf_info` helper.
- `layers` parameter (we might need to touch base on the expected behavior here)
- A quick test for the v3 `Assay` getter/setter (this can be simple, not a lot of edge cases to worry about)

**Comment by longmanz** [comment]
To-Do from discussion with David:  need to implement the same strategy used by FindVariableFeatures() which does search and match for "layer" based on the "method" being specified (when "method" is set and "layer" is NULL). 

1. copy codes from FindVariableFeatures() and modify them
2. if "method" is specified and "layer" is not, run the codes from 1). 
3. if both "method" and "layer" are specified, then do the standard check

**Comment by dcollins15** [comment]
Closed in favor of:
- https://github.com/satijalab/seurat-object/pull/245
